File: walker/walker.py
#! /usr/bin/python3

# A program that uses the os.walk module to 'walk' through files and folders in
# and including the CWD. Currently can search and delete.

import sys
import os

def search_base(keyword_list):
    # Initializes a list that will store the file paths of files to be deleted
    file_match_list = []

    # Search in and under CWD for files with all keyword in their name, then
    # append to file_match_list
    for root, dirs, files in os.walk(os.getcwd()):
        for file in files:
            if all(i in file for i in keyword_list):
                file_match_list.append(os.path.join(root, file))

    return file_match_list


def search(keyword_list):
    search_result = search_base(keyword_list)

    if search_result == []:
        print("No matches found")
    else:
        for i in search_result: print(i)

def delete_files(keyword_list):
    file_match_list = search_base(keyword_list)
    
    # Prompts user to confirm delete as long as matches have been found
    if file_match_list == []:
        print("No matches found")
    else:
        for i in file_match_list: print(i)
        user_choice = input("Are you sure you wish to delete these " + str(len(file_match_list)) + " files? [y/N]: ")

        if user_choice in "yY":
            for file in file_match_list:
                os.remove(file)
            print(str(len(file_match_list)) + " files deleted")
        elif user_choice in "nN":
            print("No files deleted")



if __name__ == '__main__':
    def initiate():
        option = input("Press 1 if you want to search, 2 if you want to search and delete: ")

        if option == "1":
            keyword_list = str.split(input("Enter a keyword (or two): "))
            search(keyword_list)
        elif option == "2":
            keyword_list = str.split(input("Enter a keyword (or two): "))
            delete_files(keyword_list)
        else:
            print("You must select an option")
            initiate()
    
    initiate()
            

File: walker/Walker.java
// A clone of walker.py, but written in Java

import java.io.*;
import java.util.*;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;

public class Walker { // Prompts user to search or delete, gets keywords and passes them to the correct method
    public static void main(String[] args) throws IOException {
        while (true) {
            System.out.print("Press 1 if you want to search, 2 if you want to search and delete: ");
            BufferedReader read_keywords = new BufferedReader(new InputStreamReader(System.in));

            try {
                int choice = Integer.parseInt(read_keywords.readLine().trim());
                if (choice != 1 && choice != 2) {
                    continue;
                }
                System.out.print("Enter one or more keywords: ");
                String[] keywords = read_keywords.readLine().trim().split("\\s+");
                
                if (choice == 1) {
                    search(keywords);
                    break;
                } else if (choice ==2) {
                    delete(keywords);
                    break;
                } 
            } catch (NumberFormatException e) {
                System.out.println("Invalid option \"" + e + "\"");
                continue;
            }
         }
    }

    public static String[] getAllFilePathsInCWD() throws IOException {
        final ArrayList<String> file_list = new ArrayList<String>();

        try { // Get all files in CWD and under and appends them to file_list
            Path startPath = Paths.get(System.getProperty("user.dir"));
            Files.walkFileTree(startPath, new SimpleFileVisitor<Path>() {
                    @Override
                        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
                        file_list.add(file.toString());
                        return FileVisitResult.CONTINUE;
                    }
                    @Override
                        public FileVisitResult visitFileFailed(Path file, IOException e) {
                        return FileVisitResult.CONTINUE;
                    }
                });
        } catch (IOException e) {
            e.printStackTrace();
            throw new IOException("File walk failed");
        }
        return file_list.toArray(new String[0]);
    }

    public static String[] getMatchingFilePaths(String[] pFilePaths, String[] pMatchList) {
        // ArrayList to store the matches
        ArrayList<String> matchingFilePaths = new ArrayList<String>();

        for (String filePath : pFilePaths) { // Match against all keywords
            boolean isMatch = true;
            String fileNameWithoutExtn = filePath.substring(filePath.lastIndexOf(File.separator) + 1);

            for (String stringToMatch : pMatchList) {
                if (!fileNameWithoutExtn.contains(stringToMatch)) {
                    isMatch = false;
                    break;
                }
            }
            if (isMatch) {
                matchingFilePaths.add(filePath);
            }
        }
        return matchingFilePaths.toArray(new String[0]);
    }

    public static void search(String[] pMatchList) throws IOException {
        String[] matches = getMatchingFilePaths(getAllFilePathsInCWD(), pMatchList);

        if (matches.length == 0) {
            System.out.println("No matches found");
            System.exit(0);
        }
        for (String i : matches) {
            System.out.println(i);
        }

        if (matches.length == 1) {
            System.out.println("Found 1 match");
        } else {
            System.out.println("Found " + matches.length + " matches");
        }
    }

    public static void delete(String[] pMatchList) throws IOException {
        BufferedReader read_choice = new BufferedReader(new InputStreamReader(System.in));
        String[] matchedFilePaths = getMatchingFilePaths(getAllFilePathsInCWD(), pMatchList);

        if (matchedFilePaths.length == 0) {
            System.out.println("No matches found");
            System.exit(0);
        }

        for (String i : matchedFilePaths) {
            System.out.println(i);
        }

        while (true) {
            if (matchedFilePaths.length == 1) {
                System.out.print("Are you sure you want to delete this file? [y/N]: ");
            } else {
                System.out.print("Are you sure you want to delete these " + matchedFilePaths.length + " files? [y/N]: ");
            }
            String choice = read_choice.readLine().trim();

            if ("yY".contains(choice)) {
                for (String filePath : matchedFilePaths) {
                    File file = new File(filePath);
                    file.delete();
                }
                System.out.println("Deleted " + matchedFilePaths.length + " files");
                System.exit(0);
            } else if ("nN".contains(choice)) {
                System.out.println("No files deleted");
                System.exit(0);
            } else {
                System.out.println("Invalid option");
            }
        }
    }
}
